From 89c25444ead2ee32efb47c068b9c34ecb2fbf09b Mon Sep 17 00:00:00 2001
From: Andreas Monzner <andreas.monzner@dream-property.net>
Date: Wed, 27 Feb 2013 19:37:42 +0100
Subject: [PATCH 2/2] Dreambox: add new flash map for 1GB based dreamboxes
 (needs reflash with new ubi(fs) based nfi), use legacy
 flash map when jffs2 is detected in the kernel cmdline
 as fallback

---
 arch/mips/brcmstb/board.c |   18 +++++++++++++++++-
 arch/mips/brcmstb/prom.c  |    9 +++++++++
 2 files changed, 26 insertions(+), 1 deletions(-)

diff --git a/arch/mips/brcmstb/board.c b/arch/mips/brcmstb/board.c
index 53d24e9..68b6b57 100644
--- a/arch/mips/brcmstb/board.c
+++ b/arch/mips/brcmstb/board.c
@@ -731,11 +731,21 @@ static struct mtd_partition fixed_partition_map[] = {
 
 #elif defined(CONFIG_DREAMBOX_DM7020HD)
 
-static struct mtd_partition fixed_partition_map[] = {
+int legacy_flash_map = 0;
+
+static struct mtd_partition fixed_partition_map_legacy[] __maybe_unused = {
 	{ .name = "complete",		.offset = 0,			.size = 1024 * 1024 * 1024,	},
 	{ .name = "loader",		.offset = 0,			.size = 1024 * 1024,		},
 	{ .name = "boot partition",	.offset = 1024 * 1024,		.size = 7 * 1024 * 1024,	},
 	{ .name = "root partition",	.offset = 8 * 1024 * 1024,	.size = 248 * 1024 * 1024,	},
+	{ .name = "misc partition",	.offset = 256 * 1024 * 1024,	.size = 768 * 1024 * 1024,	},
+};
+
+static struct mtd_partition fixed_partition_map[] __maybe_unused = {
+	{ .name = "complete",		.offset = 0,			.size = 1024 * 1024 * 1024,	},
+	{ .name = "loader",		.offset = 0,			.size = 1024 * 1024,		},
+	{ .name = "boot",		.offset = 1024 * 1024,		.size = 7 * 1024 * 1024,	},
+	{ .name = "root",		.offset = 8 * 1024 * 1024,	.size = 1016 * 1024 * 1024,	},
 };
 
 #else
@@ -771,6 +781,12 @@ static struct mtd_partition fixed_partition_map[] = {
  */
 int __init board_get_partition_map(struct mtd_partition **p)
 {
+#if defined(CONFIG_DREAMBOX_DM7020HD)
+	if (legacy_flash_map) {
+		*p = fixed_partition_map_legacy;
+		return ARRAY_SIZE(fixed_partition_map_legacy);
+	}
+#endif
 	*p = fixed_partition_map;
 	return ARRAY_SIZE(fixed_partition_map);
 }
diff --git a/arch/mips/brcmstb/prom.c b/arch/mips/brcmstb/prom.c
index 4482372..7d67450 100644
--- a/arch/mips/brcmstb/prom.c
+++ b/arch/mips/brcmstb/prom.c
@@ -459,6 +459,15 @@ void __init prom_init(void)
 		brcm_primary_macaddr[3] = (pm0DataLo & 0x00FF0000) >> 16;
 		brcm_primary_macaddr[4] = (pm0DataLo & 0x0000FF00) >> 8;
 		brcm_primary_macaddr[5] = (pm0DataLo & 0x000000FF);
+
+#if defined(CONFIG_DREAMBOX_DM7020HD)
+		if (strstr(arcs_cmdline, "jffs2")) {
+			extern int legacy_flash_map;
+			printk(KERN_WARNING "JFFS2 detected... use legacy flashmap\n");
+			legacy_flash_map = 1;
+		}
+#endif
+
 	} while(0);
 #endif
 
-- 
1.7.7

